### DISCUSSION TITLE
Redundant handler and mapper registrations in consumer projects

### DISCUSSION THREAD

**Discussion by nesbo** [BrighterCommand/Brighter#2395] [Q&A] [answered]
Hi guys!

I obviously have problem understanding the registrations of async handlers and mappers when I have IHostedService that has to build and run the Dispatcher.

So, I have regular service registration in my Worker console project, building IServiceCollection. Here I must .AddBrighter() and register async handlers and mappers explicitly (or use AutoFromAssemblies - which I'd like to avoid btw).

Then, in my IHostedService I'm building up the Dispatcher with CommandProcessorFactory, passing the SubscriberRegistry where I had to register my handlers again, and to the DispatcherBuilder passing another MessageMapper with same map like I had in the service registration in the beginning.

Only if I do the registrations in both places (or use AutoFromAssemblies), the worker starts consuming messages from the bus and instantiating handlers.

I'd really like to avoid redundant registrations if possible.

Please let me know what I'm doing wrong here.

Thanks!
Aleks

**Comment by iancooper** [comment]
@preardon here ^^^^

**Comment by iancooper** [comment]
@nesbo This is a problem with the way that we build the underlying Command Processor and Service Activator. Both are doing registration of these items. It's a bug, and I am going to raise it as an issue.

**Reply by nesbo** [reply]
thanks a lot @iancooper ! good to know. I'll see if I get some time to take a look at it deeper.

**Comment by holytshirt** [comment]
Coming back to this I think @nesbo  just needs use the dispatcher from the servicecollection in his Ihostedservice, like how we do in our extension 
https://github.com/BrighterCommand/Brighter/blob/master/src/Paramore.Brighter.ServiceActivator.Extensions.Hosting/ServiceActivatorHostedService.cs

**Reply by holytshirt** [reply, marked as answer]
and you can just use what we built like this 
https://github.com/BrighterCommand/Brighter/blob/d5c946ec7a947e8491c15794512f6881013f14d1/samples/RedisTaskQueue/GreetingsReceiver/Program.cs#L55
